tests/Unit/Evolution/Systems/Ccz4/Test_Ricci.cpp

```diff
@@ -0,0 +1,192 @@
+      }
+    }
+  }
+
+  const auto field_d_up = gr::deriv_inverse_spatial_metric(
+      inverse_conformal_spatial_metric, d_spatial_metric);
+
+  using field_d_tag =
+      Ccz4::Tags::FieldD<SpatialDim, Frame::Inertial, DataVector>;
+  Variables<tmpl::list<field_d_tag>> field_d_var(num_points_3d);
+  get<field_d_tag>(field_d_var) = field_d;
+  const auto d_field_d_var = partial_derivatives<tmpl::list<field_d_tag>>(
+      field_d_var, mesh, coord_map.inv_jacobian(x_logical));
+  const auto& d_field_d =
+      get<Tags::deriv<field_d_tag, tmpl::size_t<SpatialDim>, Frame::Inertial>>(
+          d_field_d_var);
+
+  const auto d_conformal_christoffel_second_kind =
+      Ccz4::deriv_conformal_christoffel_second_kind(
+          inverse_conformal_spatial_metric, field_d, d_field_d, field_d_up);
+
+  using conformal_factor_tag = Ccz4::Tags::ConformalFactor<DataVector>;
+  Variables<tmpl::list<conformal_factor_tag>> conformal_factor_var(
+      num_points_3d);
+  get(get<conformal_factor_tag>(conformal_factor_var)) = conformal_factor;
+  const auto d_conformal_factor_var =
+      partial_derivatives<tmpl::list<conformal_factor_tag>>(
+          conformal_factor_var, mesh, coord_map.inv_jacobian(x_logical));
+  const auto& d_conformal_factor =
+      get<Tags::deriv<conformal_factor_tag, tmpl::size_t<SpatialDim>,
+                      Frame::Inertial>>(d_conformal_factor_var);
+
+  tnsr::i<DataVector, SpatialDim, Frame::Inertial> field_p{};
+  for (size_t i = 0; i < SpatialDim; i++) {
+    field_p.get(i) = d_conformal_factor.get(i) / conformal_factor;
+  }
+
+  using field_p_tag =
+      Ccz4::Tags::FieldP<SpatialDim, Frame::Inertial, DataVector>;
+  Variables<tmpl::list<field_p_tag>> field_p_var(num_points_3d);
+  get<field_p_tag>(field_p_var) = field_p;
+  const auto d_field_p_var = partial_derivatives<tmpl::list<field_p_tag>>(
+      field_p_var, mesh, coord_map.inv_jacobian(x_logical));
+  const auto& d_field_p =
+      get<Tags::deriv<field_p_tag, tmpl::size_t<SpatialDim>, Frame::Inertial>>(
+          d_field_p_var);
+
+  const auto christoffel_second_kind = Ccz4::conformal_christoffel_second_kind(
+      inverse_conformal_spatial_metric, field_d);
+
+  // Compute expected and actual ricci tensors using above computed arguments
+  const auto expected_ricci_tensor{
+      pypp::call<tnsr::ii<DataVector, SpatialDim, Frame::Inertial>>(
+          "Ricci", "spatial_ricci_tensor", christoffel_second_kind,
+          d_conformal_christoffel_second_kind, conformal_spatial_metric,
+          inverse_conformal_spatial_metric, field_d, field_d_up, field_p,
+          d_field_p)};
+
+  const auto actual_ricci_tensor = Ccz4::spatial_ricci_tensor(
+      christoffel_second_kind, d_conformal_christoffel_second_kind,
+      conformal_spatial_metric, inverse_conformal_spatial_metric, field_d,
+      field_d_up, field_p, d_field_p);
+
+  CHECK_ITERABLE_APPROX(expected_ricci_tensor, actual_ricci_tensor);
```

Comment:
can you also compare to what you get when you call `gr::ricci_tensor`? That provides another independent check. Currently if you have the same error in your python and C++ code the test will still pass